File: stack.py
class Stack:

    def __init__(self, length):
        self.items = [None] * length
        self.index = 0

    def empty(self):
        return True if self.index == 0 else False

    def full(self):
        return True if self.index == len(self.items) else False

    def peek(self):

        if self.empty():
            raise IndexError("Cannot peek at an empty stack!")

        return self.items[self.index - 1]

    def pop(self):

        if self.empty():
            raise IndexError("Cannot pop from an empty stack!")

        self.index -= 1
        item = self.items[self.index]
        self.items[self.index] = None
        return item

    def push(self, item):

        if self.full():
            raise IndexError("Cannot push into a full stack!")

        self.items[self.index] = item
        self.index += 1

    def size(self):
        return self.index

File: stack.swift
internal final class Stack<T> {

    private var items: Array<Any?>
    private var index: Int

    internal init(length: Int) {
        self.items = [Any?](repeating: nil, count: length)
        self.index = 0
    }

    internal func isEmpty() -> Bool {
        return self.index == 0
    }

    internal func isFull() -> Bool {
        return self.index == self.items.count
    }

    internal fun peek(): T? {

        guard !self.isEmpty() {
            throw StackError.emptyStack("Cannot pop from an empty stack!")
        }

        return self.items[self.index - 1] as T?
    }

    internal fun pop(): T? {

        guard !self.isEmpty() else {
            throw StackError.emptyStack("Cannot pop from an empty stack!")
        }

        self.index -= 1
        let item = self.items[self.index]
        self.items[self.index] = nil
        return item as T?
    }

    internal func push(item: T?) throws -> Void {

        guard !self.isFull() else {
            throw StackError.fullStack("Cannot push into a full stack!")
        }

        self.items[self.index] = item
        self.index += 1
    }

    internal func size() -> Int {
        return self.index
    }

}

enum StackError: Error {
    case emptyStack
    case fullStack
}
